File: java/ex-011-xxe.java
// Dangerous: XML External Entity (XXE)
// Run: javac ex-011-xxe.java && java XXEDemo "<!DOCTYPE foo [<!ENTITY xxe SYSTEM 'file:///etc/hosts'>]><a>&xxe;</a>"
// Ref: OWASP XXE
import javax.xml.parsers.*;
import org.w3c.dom.*;
import java.io.*;

public class XXEDemo {
    public static void main(String[] args) throws Exception {
        String xml = args.length>0? args[0] : "<a>ok</a>";
        DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
        // VULN: features left default (DTDs enabled on some impls)
        DocumentBuilder db = dbf.newDocumentBuilder();
        Document doc = db.parse(new ByteArrayInputStream(xml.getBytes("UTF-8")));
        System.out.println(doc.getDocumentElement().getTextContent());
    }
}

File: csharp/ex-011-xxe.cs
// Dangerous: XXE in .NET XmlDocument
// Ref: Microsoft secure XML processing
using System;
using System.Xml;

class XXEDemoCs {
    static void Main(string[] args) {
        string xml = args.Length>0 ? args[0] : "<a>ok</a>";
        var doc = new XmlDocument();
        // VULN: default settings may resolve external entities
        doc.XmlResolver = new XmlUrlResolver();
        doc.LoadXml(xml);
        Console.WriteLine(doc.DocumentElement.InnerText);
    }
}
